Doesn't work with `impl Trait` return type

A minimal use-case of what I'd need:
```rust
use std::io;

struct Foo;

impl Foo {
    #[no_panic::no_panic]
    fn get(&mut self) -> io::Result<impl io::Write> {
        Ok(Vec::new())
    }
}
```
Error:
```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
 --> src\lib.rs:7:37
  |
7 |     fn get(&mut self) -> io::Result<impl io::Write> {
  |                                     ^^^^^^^^^^^^^^
```
A minimal reproduction case:
```rust
#[no_panic::no_panic]
fn get() -> impl std::io::Write {
    Vec::new()
}
```
Error:
```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
 --> src\lib.rs:2:17
  |
2 |     fn get() -> impl std::io::Write {
  |                 ^^^^^^^^^^^^^^^^^^^
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -127,7 +127,10 @@ use proc_macro::TokenStream;
 use proc_macro2::{Span, TokenStream as TokenStream2};
 use quote::quote;
 use syn::parse::{Error, Nothing, Result};
-use syn::{parse_quote, Attribute, FnArg, Ident, ItemFn, Pat, PatType, ReturnType};
+use syn::{
+    parse_quote, Attribute, FnArg, GenericArgument, Ident, ItemFn, Pat, PatType, Path,
+    PathArguments, ReturnType, Token, Type, TypeInfer, TypeParamBound,
+};
 
 #[proc_macro_attribute]
 pub fn no_panic(args: TokenStream, input: TokenStream) -> TokenStream {
@@ -153,6 +156,45 @@ fn parse(args: TokenStream2, input: TokenStream2) -> Result<ItemFn> {
         ));
     }
     Ok(function)
+}
+
+// Convert `Path<impl Trait>` to `Path<_>`
+fn make_impl_trait_wild(ret: &mut Type) {
+    match ret {
+        Type::ImplTrait(impl_trait) => {
+            *ret = Type::Infer(TypeInfer {
+                underscore_token: Token![_](impl_trait.impl_token.span),
+            });
+        }
+        Type::Array(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::Group(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::Paren(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::Path(ret) => make_impl_trait_wild_in_path(&mut ret.path),
+        Type::Ptr(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::Reference(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::Slice(ret) => make_impl_trait_wild(&mut ret.elem),
+        Type::TraitObject(ret) => {
+            for bound in &mut ret.bounds {
+                if let TypeParamBound::Trait(bound) = bound {
+                    make_impl_trait_wild_in_path(&mut bound.path);
+                }
+            }
+        }
+        Type::Tuple(ret) => ret.elems.iter_mut().for_each(make_impl_trait_wild),
+        _ => {}
+    }
+}
+
+fn make_impl_trait_wild_in_path(path: &mut Path) {
+    for segment in &mut path.segments {
+        if let PathArguments::AngleBracketed(bracketed) = &mut segment.arguments {
+            for arg in &mut bracketed.args {
+                if let GenericArgument::Type(arg) = arg {
+                    make_impl_trait_wild(arg);
+                }
+            }
+        }
+    }
 }
 
 fn expand_no_panic(mut function: ItemFn) -> TokenStream2 {
@@ -197,7 +239,11 @@ fn expand_no_panic(mut function: ItemFn) -> TokenStream2 {
 
     let ret = match &function.sig.output {
         ReturnType::Default => quote!(-> ()),
-        output @ ReturnType::Type(..) => quote!(#output),
+        ReturnType::Type(arrow, output) => {
+            let mut output = output.clone();
+            make_impl_trait_wild(&mut output);
+            quote!(#arrow #output)
+        }
     };
     let stmts = function.block.stmts;
     let message = format!(

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -202,6 +202,17 @@ assert_no_panic! {
 
         fn main() {}
     }
+
+    mod test_return_impl_trait {
+        use std::io;
+
+        #[no_panic]
+        pub fn f() -> io::Result<impl io::Write> {
+            Ok(Vec::new())
+        }
+
+        fn main() {}
+    }
 }
 
 assert_link_error! {

